### PR TITLE
Detect NotNull Annotations for Java Methods

### PR DESCRIPTION
**Pull request by noti0na1**

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/Definitions.scala [hunk 1]
```diff
@@ -813,6 +813,22 @@ class Definitions {
   @tu lazy val InfixAnnot: ClassSymbol = ctx.requiredClass("scala.annotation.infix")
   @tu lazy val AlphaAnnot: ClassSymbol = ctx.requiredClass("scala.annotation.alpha")
 
+  @tu lazy val NotNullAnnots: List[ClassSymbol] =
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/Definitions.scala:819]: Could you please add a comment explaining what these are?

File: compiler/src/dotty/tools/dotc/core/Definitions.scala [hunk 2]
```diff
@@ -813,6 +813,22 @@ class Definitions {
...
+      "android.annotation.NonNull" ::
+      "com.android.annotations.NonNull" ::
+      "org.eclipse.jdt.annotation.NonNull" ::
+      "org.checkerframework.checker.nullness.qual.NonNull" ::
+      "org.checkerframework.checker.nullness.compatqual.NonNullDecl" ::
+      "org.jetbrains.annotations.NotNull" ::
+      "lombok.NonNull" ::
+      "io.reactivex.annotations.NonNull" ::
+      // A special NotNull annotation for test
+      "mytests.annot.TestNotNull" :: Nil).map(s => ctx.requiredClass(PreNamedString(s)))
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/Definitions.scala]: We will need to delete this one before merging. We should write the test using one of the "standard" annotations (probably javax.annotation.Nonnull).

Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/Definitions.scala]: For my knowledge, why do we need `PreNamedString` here? Why not just use `s` directly?

File: compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala [hunk 3]
```diff
@@ -60,14 +60,18 @@ object JavaNullInterop {
     if (sym.name == nme.TYPE_ || sym.isAllOf(Flags.JavaEnumValue))
       // Don't nullify the `TYPE` field in every class and Java enum instances
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala:61]: Could you please modify this comment to read

// Don't nullify the `TYPE` field in every class.
// Don't nullify Java enum instances.

File: compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala [hunk 4]
```diff
@@ -60,14 +60,18 @@ object JavaNullInterop {
...
       tp
-    else if (sym.name == nme.toString_ || sym.isConstructor)
+    else if (sym.name == nme.toString_ || sym.isConstructor || hasNotNull(sym))
       // Don't nullify the return type of the `toString` method and constructors
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala]: Similarly, could you please break this comment up into multiple lines?

i.e.

// Don't nullify the return type of the `toString` method.
// Don't nullify the return type of constructors.
// Don't nullify the return type of methods with a not-null annotation.

File: compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala [hunk 5]
```diff
@@ -60,14 +60,18 @@ object JavaNullInterop {
...
+      // or it has a NotNull annotation
       nullifyParamsOnly(tp)
     else
       // Otherwise, nullify everything
       nullifyType(tp)
   }
 
+  private def hasNotNull(sym: Symbol)(implicit ctx: Context): Boolean =
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala]: Let's rename this to `hasNotNullAnnot`, so that it's clear that it checks for annotations.

File: compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala [hunk 6]
```diff
@@ -60,14 +60,18 @@ object JavaNullInterop {
     if (sym.name == nme.TYPE_ || sym.isAllOf(Flags.JavaEnumValue))
       // Don't nullify the `TYPE` field in every class and Java enum instances
       tp
-    else if (sym.name == nme.toString_ || sym.isConstructor)
+    else if (sym.name == nme.toString_ || sym.isConstructor || hasNotNull(sym))
       // Don't nullify the return type of the `toString` method and constructors
+      // or it has a NotNull annotation
       nullifyParamsOnly(tp)
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala]: Will this handle fields correctly? The test seems to indicate it does, but why does it work? Do fields get desugared into getters and setters?

File: tests/explicit-nulls/pos-separate/notnull/J_1.java [hunk 7]
```diff
@@ -0,0 +1,21 @@
+package mytests.annot;
+
+public class J_1 {
+
+    @TestNotNull
+    public static final String s = "s";
```
Comment (abeln) [inline tests/explicit-nulls/pos-separate/notnull/J_1.java]: We also need to test an instance (non-static) field that's annotated as non-null.

File: tests/explicit-nulls/pos-separate/notnull/J_1.java [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+        return "m: " + i;
+    }
+
+    @TestNotNull
+    public static String n(int i) {
+        return "n: " + i;
+    }
+}
+
+@interface TestNotNull {
```
Comment (abeln) [inline tests/explicit-nulls/pos-separate/notnull/J_1.java]: Would it be possible to use a standard annotation? We shouldn't really add a dummy annotation to our whitelist just so that this test works.

**Review by abeln — CHANGES_REQUESTED** [review]
Nice! Thanks for looking into this. I've added a few comments.

**Comment by noti0na1** [comment]
@abeln I just find, in ClassFileParser, if the value of a final field is not null, the field will have a ConstantType; but in JavaParser, a field will never get a ConstantType.

**Comment by noti0na1** [comment]
https://gitter.im/lampepfl/dotty?at=5db0bae3e886fb5aa2e2b61d
> @noti0na1 it looks like this was added in 2016 to the JavaParser in scalac, so it's probably just a matter of porting that: [scala/scala@81d2c61](https://github.com/scala/scala/commit/81d2c6185e6c10defcf1f6a25f69e8b2ed025d86)
>though that only supports literals, more complex expressions would require a bit more work on the parser: [scala/bug#10410](https://github.com/scala/bug/issues/10410)

File: compiler/src/dotty/tools/dotc/core/Definitions.scala (hunk 1 above)
Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/core/Definitions.scala:819]: Done

File: compiler/src/dotty/tools/dotc/core/Definitions.scala (hunk 2 above)
Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/core/Definitions.scala]: PreNamedString is removed

File: compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala (hunk 3 above)
Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala:61]: Done

File: compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala (hunk 5 above)
Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala]: I will push this in the next commit

File: compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala (hunk 6 above)
Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala]: The logic of `JavaNullMap` is updated

File: tests/explicit-nulls/pos-separate/notnull/J_1.java (hunk 7 above)
Comment (noti0na1) [reply tests/explicit-nulls/pos-separate/notnull/J_1.java]: Tests added

File: tests/explicit-nulls/pos-separate/notnull/J_1.java (hunk 8 above)
Comment (noti0na1) [reply tests/explicit-nulls/pos-separate/notnull/J_1.java]: dummy annotation is removed

File: compiler/src/dotty/tools/dotc/core/Definitions.scala [hunk 9]
```diff
@@ -813,6 +813,21 @@ class Definitions {
   @tu lazy val InfixAnnot: ClassSymbol = ctx.requiredClass("scala.annotation.infix")
   @tu lazy val AlphaAnnot: ClassSymbol = ctx.requiredClass("scala.annotation.alpha")
 
+  // A list of NotNull annotations
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/Definitions.scala]: We should try to leave comments that provide context about how the code is used.
In this case, I would say

// A list of annotations that are commonly used to indicate that a field/method argument or return type is
// not null.
// These annotations are used by the nullification logic in `JavaNullInterop` to improve the precision of type nullification.

File: compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala [hunk 10]
```diff
@@ -60,35 +60,45 @@ object JavaNullInterop {
...
-  private def nullifyParamsOnly(tp: Type)(implicit ctx: Context): Type =
-    new JavaNullMap(alreadyNullable = false, nonNullResultType = true)(ctx)(tp)
+  private def hasNotNullAnnot(sym: Symbol)(implicit ctx: Context): Boolean =
+    ctx.definitions.NotNullAnnots.exists(nna => sym.unforcedAnnotation(nna).isDefined)
+
+  /** If tp is a MethodType, the parameters and the inside of return type are nullified,
+   *  but the result return type is not nullable.
+   *  If tp is a type of a field, the inside of the type is nullified,
+   *  but the result type is not nullable.
+  */
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala]: nit: fix indentation

File: compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala [hunk 11]
```diff
@@ -60,35 +60,45 @@ object JavaNullInterop {
...
 
 
   /** A type map that adds `| JavaNull`.
-   *  @param alreadyNullable whether the type being mapped is already nullable (at the outermost level).
-   *                         This is needed so that `JavaNullMap(A | B)` gives back `(A | B) | JavaNull`,
-   *                         instead of `(A|JavaNull | B|JavaNull) | JavaNull`.
-   *  @param nonNullResultType if true, then the current type is a method or generic method type,
-   *                           and we don't want to nullify its return type at the top level.
-   *                           This is useful e.g. for constructors.
+   *  If a field has a final flag and its value is not null, then the type of it is a ConstantType,
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala]: This comment belongs below line 151:

// In all other cases, return the type unchanged.
// In particular, if the type is a ConstantType, then we don't nullify it because it is the type of a final non-nullable field.
case _ => tp

File: compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala [hunk 12]
```diff
@@ -60,35 +60,45 @@ object JavaNullInterop {
...
+   *  we don't need to nullify it.
+   *  @param currentLevelNullable whether the type needs to be nullable at current level.
+   *                         Initialize this value to true, so that the field types and the return types of
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala]: nit: fix indentation

File: tests/explicit-nulls/pos-separate/notnull/J_2.java [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+        return "g: " + i;
+    }
+
+    @Nonnull
+    public String h(int i) {
+        return "h: " + i;
+    }
+
+    @Nonnull
+    public <T> String[] gemericf(T a) {
```
Comment (abeln) [inline tests/explicit-nulls/pos-separate/notnull/J_2.java]: gemericf -> genericf

File: tests/explicit-nulls/pos-separate/notnull/J_2.java [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+        String[] as = new String[1];
+        as[0] = "" + a;
+        return as;
+    }
+
+    @Nonnull
+    public <T> List<T> gemericg(T a) {
```
Comment (abeln) [inline tests/explicit-nulls/pos-separate/notnull/J_2.java]: gemericg -> genericg

File: tests/explicit-nulls/pos-separate/notnull/S_3.scala [hunk 15]
```diff
@@ -0,0 +1,15 @@
+// Test the NotNull annotations are working in class files'
```
Comment (abeln) [inline tests/explicit-nulls/pos-separate/notnull/S_3.scala]: remote stray ' at the end of the line

File: tests/explicit-nulls/pos/notnull/S.scala [hunk 16]
```diff
@@ -0,0 +1,15 @@
+// Test the NotNull annotations are working in class files'
```
Comment (abeln) [inline tests/explicit-nulls/pos/notnull/S.scala]: This test exercises the java parser, not the class file, right?
If so, let's change the comment.

**Review by abeln — CHANGES_REQUESTED** [review]
It's looking good! I left a few more comments.

**Comment by abeln** [comment]
It's very nice to see this working!

File: compiler/src/dotty/tools/dotc/core/Definitions.scala (hunk 9 above)
Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/core/Definitions.scala]: Comment added

File: compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala (hunk 10 above)
Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala]: Done

File: compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala (hunk 11 above)
Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala]: Comment added

File: compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala (hunk 12 above)
Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala]: Done

File: tests/explicit-nulls/pos-separate/notnull/J_2.java (hunk 13 above)
Comment (noti0na1) [reply tests/explicit-nulls/pos-separate/notnull/J_2.java]: Done

File: tests/explicit-nulls/pos-separate/notnull/J_2.java (hunk 14 above)
Comment (noti0na1) [reply tests/explicit-nulls/pos-separate/notnull/J_2.java]: Done

File: tests/explicit-nulls/pos-separate/notnull/S_3.scala (hunk 15 above)
Comment (noti0na1) [reply tests/explicit-nulls/pos-separate/notnull/S_3.scala]: Removed

File: tests/explicit-nulls/pos/notnull/S.scala (hunk 16 above)
Comment (noti0na1) [reply tests/explicit-nulls/pos/notnull/S.scala]: Done

**Comment by noti0na1** [comment]
@abeln Updated

File: compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala [hunk 17]
```diff
@@ -60,35 +60,43 @@ object JavaNullInterop {
...
 
 
   /** A type map that adds `| JavaNull`.
-   *  @param alreadyNullable whether the type being mapped is already nullable (at the outermost level).
-   *                         This is needed so that `JavaNullMap(A | B)` gives back `(A | B) | JavaNull`,
-   *                         instead of `(A|JavaNull | B|JavaNull) | JavaNull`.
-   *  @param nonNullResultType if true, then the current type is a method or generic method type,
-   *                           and we don't want to nullify its return type at the top level.
-   *                           This is useful e.g. for constructors.
+   *  @param currentLevelNullable whether the type needs to be nullable at current level.
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala]: Hmm, this comments seems wrong, because if `currentLevelNullable = true`, then the type will _not_ be nullified at the outermost level.

I propose we rename this parameter to `outermostLevelAlreadyNullable`, and change the comment to

```Scala
/*
A type map that implements the nullification function on types. Given a Java-sourced type, this adds `| JavaNull`
in the right places to make the nulls explicit in Scala.

@param outermostLevelAlreadyNullable whether this type is already nullable at the outermost level.
                                                                 For example, `Array[String]|JavaNull` is already nullable at the outermost 
                                                                 level, but `Array[String|JavaNull]` isn't.
                                                                 If this parameter is set to true, then the types of fields, and the return types
                                                                 of methods will not be nullified.
                                                                 This is useful for e.g. constructors, and also so that `A & B` is nullified to 
                                                                 `(A & B) | JavaNull`, instead of `(A|JavaNull & B|JavaNull) | JavaNull`.
*/

File: tests/explicit-nulls/pos/notnull/S.scala [hunk 18]
```diff
@@ -0,0 +1,15 @@
+// Test the NotNull annotations are working in Java files
```
Comment (abeln) [inline tests/explicit-nulls/pos/notnull/S.scala]: the -> that

nit: missing period

File: tests/explicit-nulls/pos-separate/notnull/S_3.scala [hunk 19]
```diff
@@ -0,0 +1,15 @@
+// Test the NotNull annotations are working in class files
```
Comment (abeln) [inline tests/explicit-nulls/pos-separate/notnull/S_3.scala]: nit: missing period

**Review by abeln — CHANGES_REQUESTED** [review]
I have one comment about changing a comment (a meta-comment :-) ).

File: compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala [hunk 20]
```diff
@@ -60,35 +60,46 @@ object JavaNullInterop {
...
+   *                                       types of methods will not be nullified.
+   *                                       This is useful for e.g. constructors, and also so that `A & B` is nullified
+   *                                       to `(A & B) | JavaNull`, instead of `(A|JavaNull & B|JavaNull) | JavaNull`.
    */
-  private class JavaNullMap(var alreadyNullable: Boolean, var nonNullResultType: Boolean)(implicit ctx: Context) extends TypeMap {
-    /** Should we nullify `tp` at the outermost level? */
-    def needsTopLevelNull(tp: Type): Boolean =
-      !alreadyNullable && (tp match {
+  private class JavaNullMap(var outermostLevelAlreadyNullable: Boolean)(implicit ctx: Context) extends TypeMap {
+    /** Should we nullify `tp` at the current level? */
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala]: Why this change?

File: compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala [hunk 21]
```diff
@@ -116,31 +127,32 @@ object JavaNullInterop {
       def toJavaNullableUnion(tpe: Type): Type = OrType(tpe, defn.JavaNullAliasType)
 
       tp match {
-        case tp: TypeRef if needsTopLevelNull(tp) => toJavaNullableUnion(tp)
+        case tp: TypeRef if needsNull(tp) => toJavaNullableUnion(tp)
         case appTp @ AppliedType(tycons, targs) =>
-          val oldAN = alreadyNullable
-          alreadyNullable = false
+          val oldCLN = outermostLevelAlreadyNullable
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala]: Let's update oldCLN to oldOutermostNullable?

**Review by abeln — APPROVED** [review]
Feel free to merge after addressing the comments. Thanks!
